RUM-7018: RN Image Recording Support

marco-saia-datadog opened the pull request:
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

jonathanmos reviewed (commented):
For the `drawOnCanvas` copying of the drawables, it should be safe as the only place that references it in the call chain looks like it already copies the drawable. The methods that are protected I think it's enough indication that they are internal. Regarding the `ImageViewUtils`, if we can move it to the internal module and you can access it from there it will probably be the best solution

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -526,7 +536,26 @@ internal class DefaultImageWireframeHelperTest {
 
     @Test
-    fun `M return null W createImageWireframeByDrawable { application context is null }`() {
+    fun `M not return image wireframe W createImageWireframe { ImagePrivacy MASK_LARGE_ONLY & isContextual image}`() {
```
```suggestion
    fun `M not return image wireframe W createImageWireframeByBitmap { ImagePrivacy MASK_LARGE_ONLY & isContextual image}`() {
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -503,7 +494,26 @@ internal class DefaultImageWireframeHelperTest {
+    }
+
+    @Test
+    fun `M not return image wireframe W createImageWireframe(usePIIPlaceholder = true) { ImagePrivacy MASK_NONE }`() {
```
```suggestion
    fun `M not return image wireframe W createImageWireframeByDrawable(usePIIPlaceholder = true) { ImagePrivacy MASK_NONE }`() {
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -526,7 +536,26 @@ internal class DefaultImageWireframeHelperTest {
+    }
+
+    @Test
+    fun `M return null W createImageWireframe() { application context is null }`() {
```
```suggestion
    fun `M return null W createImageWireframeByDrawable() { application context is null }`() {
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -761,7 +790,200 @@ internal class DefaultImageWireframeHelperTest {
+            customResourceIdCacheKey = null,
+            asyncJobStatusCallback = mockAsyncJobStatusCallback
+        )
+        wireframes[0] as MobileSegment.Wireframe.ImageWireframe
```
result is not stored anywhere, is it expected?

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -761,7 +790,200 @@ internal class DefaultImageWireframeHelperTest {
+        }
+        verify(mockAsyncJobStatusCallback, times(2)).jobStarted()
+        verify(mockAsyncJobStatusCallback, times(2)).jobFinished()
+        assertThat(wireframes.size).isEqualTo(2)
```
```suggestion
        assertThat(wireframes).hasSize(2)
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -809,7 +1031,7 @@ internal class DefaultImageWireframeHelperTest {
 
     @Test
-    fun `M resolve drawable width and height W createImageWireframeByDrawable { TextView }`() {
+    fun `M resolve drawable width and height W createImageWireframe() { TextView }`() {
```
```suggestion
    fun `M resolve drawable width and height W createImageWireframeByDrawable() { TextView }`() {
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -845,7 +1067,7 @@ internal class DefaultImageWireframeHelperTest {
 
     @Test
-    fun `M not try to resolve bitmap W createImageWireframeByDrawable { PII image }`(
+    fun `M not try to resolve bitmap W createImageWireframe() { PII image }`(
```
```suggestion
    fun `M not try to resolve bitmap W createImageWireframeByDrawable() { PII image }`(
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -893,7 +1115,7 @@ internal class DefaultImageWireframeHelperTest {
 
     @Test
-    fun `M try to resolve bitmap W createImageWireframeByDrawable { non-PII image }`() {
+    fun `M try to resolve bitmap W createImageWireframe() { non-PII image }`() {
```
```suggestion
    fun `M try to resolve bitmap W createImageWireframeByDrawable() { non-PII image }`() {
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/DefaultImageWireframeHelperTest.kt:
```diff
@@ -929,7 +1151,7 @@ internal class DefaultImageWireframeHelperTest {
 
     @Test
-    fun `M return content placeholder W createImageWireframeByDrawable { PII image }`() {
+    fun `M return content placeholder W createImageWireframe() { PII image }`() {
```
```suggestion
    fun `M return content placeholder W createImageWireframeByDrawable() { PII image }`() {
```

0xnm commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtilsTest.kt:
```diff
@@ -379,14 +373,19 @@ internal class DrawableUtilsTest {
     @Test
-    fun `M not use original drawable W createBitmapOfApproxSizeFromDrawable`() {
+    fun `M uses original drawable W createBitmapOfApproxSizeFromDrawable`() {
+        /**
```
this is to be removed before the merge? it is better to add it as PR comment rather than code comment.

0xnm commented on features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtils.kt:
```diff
@@ -103,27 +100,26 @@ internal class DrawableUtils(
-            // `constantState` contains only immutable properties of drawable,the state needs to be set manually
-            setState(drawable.current.state)
-        }
+        // PR NOTE:
+        // The logic to create a copy of the drawable was removed.
+        // We should verify that it is safe not to copy the drawable.
+        // Currently, this function seems to be used only with drawables that were already cloned.
```
should be PR comment, rather than code comment

marco-saia-datadog commented:
@jonathanmos

I had to do some refactoring to move the ImageViewUtils to the dd-sdk-android-internal package, because the class was relying on Session Replay `MobileSegment.WireframeClip` class. I have refactored it to make it use `Rect` instead, and added an extension function to easily convert a `Rect` to a `MobileSegment.WireframeClip`, and modified the code and tests accordingly.

jonathanmos commented on features/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/internal/utils/DrawableUtilsTest.kt:
```diff
@@ -379,14 +373,13 @@ internal class DrawableUtilsTest {
 
     @Test
-    fun `M not use original drawable W createBitmapOfApproxSizeFromDrawable`() {
+    fun `M uses original drawable W createBitmapOfApproxSizeFromDrawable`() {
```
It's not the original drawable since we now rely on the drawable being copied previously. I'm not sure this test makes sense to keep
